src/Sparrow/Utils/DiffPages.cs

```diff
@@ -20,90 +21,237 @@ public unsafe class DiffPages
+                // We write the actual size of the stored data.
+                rangePtr->Count = (long)ptr - rangePtr->Start;
+                writeCount += rangePtr->Count;
+
+                // If we are going to step over the temporary range data, just copy it full.
+                if ((byte*)originalBuffer + writeCount >= (byte*)rangePtr)
+                    goto CopyFull;
             }
             else
             {
-                WriteDiffNonZeroes(start * sizeof(long), length, (byte*)modifiedBuffer);
+                rangePtr++;
             }
 
+            byte* writePtr = Output;
+            for (Range* end = (Range*)(Output + size); rangePtr < end; rangePtr++)
+            {
+                *((Range*)writePtr) = *rangePtr;                
+                Unsafe.CopyBlock(writePtr + 16, (byte*)modifiedBuffer + rangePtr->Start, (uint)rangePtr->Count);
+
+                writePtr += rangePtr->Count + 16;
+            }            
+
+            this.OutputSize = writePtr - Output;
+            this.IsDiff = true;
+
             return;
 
             CopyFull:
             CopyFullBuffer((byte*)modifiedBuffer, size);
         }
 
+        private void WriteDiff(byte* start, byte* end, byte* srcPtr, long destOffset)
+        {
+            Debug.Assert(end - start > 0);
+            Debug.Assert((OutputSize % sizeof(long)) == 0);
+
+            byte* outputPtr = Output + OutputSize;
+
+            long startIdx = start - srcPtr; // The start index of the run based from the start of the page we are diffing.
+            long runLengthInBytes = end - start; // The run length from the start to the end.
+
+            *(long*)(outputPtr + 0) = startIdx;
+            *(long*)(outputPtr + 8) = runLengthInBytes;
+
+            Unsafe.CopyBlock(outputPtr + 16, srcPtr + (destOffset + startIdx), (uint)runLengthInBytes);
+
+            OutputSize += runLengthInBytes + 16;
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteDiff(ulong* start, ulong* end, ulong* srcPtr, long destOffset)
+        {
+            Debug.Assert(end - start > 0);
+            Debug.Assert((OutputSize % sizeof(long)) == 0);
+
+            long outputSize = OutputSize;
+            long* outputPtr = (long*)Output + outputSize / sizeof(long);
+
+            long startIdx = start - srcPtr; // The start index of the run based from the start of the page we are diffing.
+            long runLengthInBytes = (end - start) * sizeof(ulong); // The run length from the start to the end.
+            outputPtr[0] = startIdx * sizeof(ulong);
+            outputPtr[1] = runLengthInBytes;
+            outputSize += sizeof(long) * 2;
+
+            Unsafe.CopyBlock(Output + outputSize, (srcPtr + destOffset) + startIdx, (uint)runLengthInBytes);
+
+            OutputSize = outputSize + runLengthInBytes;
+        }
+
+        //public void ComputeDiff(void* originalBuffer, void* modifiedBuffer, int size)
```

Comment:
Can be removed?

Reply:
Mhhh my bad, will not have to modify that (that was used for testing purposes). Removed and pushed already.